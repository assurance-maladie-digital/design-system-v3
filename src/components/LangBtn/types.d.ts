export interface Language {
	direction: string
	name: string
	nativeName: string
}

export type AllLanguagesChar = '*'
