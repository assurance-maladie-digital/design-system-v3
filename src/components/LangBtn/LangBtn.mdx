import { Controls, Canvas, Meta, Source } from '@storybook/blocks';import * as LangBtnStories from './LangBtn.stories';

<Meta of={LangBtnStories} />

# LangBtn

Le composant `LangBtn` est utilisé pour afficher une alerte à l’utilisateur.

<Canvas of={LangBtnStories.Default} />

# API

<Controls of={LangBtnStories.Default} />

# Exemple d'utilisation

<Source dark code={`
<script setup lang="ts">
 import LangBtn from '@/components/LangBtn/LangBtn.vue'
 import { ref } from 'vue'

 const value = ref('fr')
 const items = ['fr', 'en', 'co']
 const ariaLabel = 'Choix de la langue'
</script>

<template>
 <div class="mt-12 ml-12">
  <LangBtn
   v-model="value"
   :available-languages="items"
   :aria-label="ariaLabel"
   hide-down-arrow
  />
 </div>
</template>
`} />