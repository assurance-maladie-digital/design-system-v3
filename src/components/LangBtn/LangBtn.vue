<script setup lang="ts">
	import { computed, ref, watch } from 'vue'
	import type { AllLanguagesChar } from '@/components/LangBtn/types'
	import { mdiMenuDown } from '@mdi/js'
	import { locales } from './locales'
	import ISO6391 from 'iso-639-1'
	import useCustomizableOptions, { type CustomizableOptions } from '@/composables/useCustomizableOptions'
	import defaultOptions from './config'

	const props = withDefaults(defineProps<CustomizableOptions & {
		modelValue?: string
		hideDownArrow?: boolean
		ariaLabel?: string
		ariaOwns?: string
		availableLanguages?: string[] | AllLanguagesChar
	}>(), {
		modelValue: 'fr',
		hideDownArrow: false,
		ariaLabel: locales.label,
		ariaOwns: 'lang-btn',
		availableLanguages: () => ['fr', 'en'],
		validator: (value: string[] | AllLanguagesChar): boolean => {
			if (Array.isArray(value)) {
				return value.length > 0
			}
			else {
				return value === '*'
			}
		},
	})

	const options = useCustomizableOptions(defaultOptions, props)

	const emit = defineEmits(['update:modelValue', 'change'])
	const menu = ref(false)
	const selectedLanguage = ref(props.modelValue)

	watch(
		() => props.modelValue,
		(newVal) => {
			selectedLanguage.value = newVal
		},
	)

	function updateLang(lang: string) {
		selectedLanguage.value = lang
		emit('update:modelValue', lang)
		emit('change', lang)
		menu.value = false
	}

	interface LanguageInfo {
		code: string
		name: string
		nativeName: string
	}

	type LanguagesData = Record<string, LanguageInfo>

	const isMenuOpen = computed(() => menu.value)
	const menuId = computed(() => `lang-menu-id`)

	const languagesData = computed<LanguagesData>(() => {
		const data: LanguagesData = {}
		let languageCodes: string[]

		if (props.availableLanguages === '*') {
			languageCodes = ISO6391.getAllCodes()
		}
		else {
			languageCodes = props.availableLanguages as string[]
		}

		languageCodes.forEach((language) => {
			data[language] = {
				code: language,
				name: ISO6391.getName(language) || language,
				nativeName: ISO6391.getNativeName(language) || ISO6391.getName(language) || language,
			}
		})

		return data
	})

	const currentLangData = computed(() => {
		const langInfo = languagesData.value[selectedLanguage.value]
		return {
			name: langInfo?.nativeName || selectedLanguage.value,
			label: `${props.ariaLabel} ${langInfo?.nativeName || selectedLanguage.value}`,
		}
	})

	defineExpose({
		currentLangData,
		updateLang,
		selectedLanguage,
	})
</script>

<template>
	<div :id="menuId">
		<VMenu
			v-bind="options.menu"
			:id="isMenuOpen ? 'lang-menu' : menuId "
			v-model="menu"
			role="menu"
			location="bottom"
		>
			<template #activator="{ props: activatorProps }">
				<VBtn
					id="lang-menu-btn"
					:aria-label="`${props.ariaLabel} ${currentLangData.name}`"
					aria-haspopup="menu"
					:aria-controls="menuId"
					:aria-owns="menuId"
					:aria-expanded="isMenuOpen"
					v-bind="{
						...options.btn,
						...activatorProps,
					}"
					class="vd-lang-btn"
				>
					{{ currentLangData.name }}
					<VIcon
						v-if="!hideDownArrow"
						v-bind="options.icon"
						class="ml-1"
					>
						{{ mdiMenuDown }}
					</VIcon>
				</VBtn>
			</template>
			<VList
				v-bind="options.list"
				aria-labelledby="lang-menu-btn"
			>
				<VListItem
					v-for="(langData, code, index) in languagesData"
					v-bind="options.listTile"
					:key="code"
					role="menuitem"
					:tabindex="index + 1"
					:aria-label="`${props.ariaLabel} ${langData.nativeName}`"
					:aria-labelledby="`${menuId} ${langData.nativeName}`"
					@click="updateLang(code)"
				>
					<VListItemTitle v-bind="options.listTileTitle">
						{{ langData.nativeName }}
					</VListItemTitle>
				</VListItem>
			</VList>
		</VMenu>
	</div>
</template>
<style lang="scss" scoped>
@use '@/assets/tokens.scss';

.v-list-item:hover {
  background-color: rgba(tokens.$colors-overlay, 0.005)
}

.vd-lang-btn {
  font-size: 16px;
  --hoverColor: rgba(tokens.$colors-overlay, 0.5);
  text-transform: none;
  letter-spacing: inherit;
}
</style>
