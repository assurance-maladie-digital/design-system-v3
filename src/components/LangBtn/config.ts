import type { VariantType } from '@/types/vuetifyTypes'

const defaultOptions = {
	menu: {
		offsetY: true,
	},
	btn: {
		color: 'primary',
		variant: 'outlined' as VariantType,
		ripple: true,
	},
	icon: {
		class: 'ml-1',
	},
}

export default defaultOptions
