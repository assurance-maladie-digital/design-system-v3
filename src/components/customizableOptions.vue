// src/components/customizableOptions.vue
<script setup>
	import useCustomizableOptions from '../composables/useCustomizableOptions'

	const defaultOptions = {
		btn: {
			color: 'primary',
		},
	}

	const props = defineProps({
		vuetifyOptions: {
			type: Object,
			default: () => ({}),
		},
	})

	const options = useCustomizableOptions(defaultOptions, props)
</script>

<template>
	<div>
		<VBtn v-bind="options.btn">
			Button
		</VBtn>
	</div>
</template>
