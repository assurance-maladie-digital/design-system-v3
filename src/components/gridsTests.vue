<script setup lang="ts">
	import { reactive, ref, onMounted } from 'vue'
	const state = reactive({
		screenWidth: 0,
		gridColumns: 1,
	})
	const calculateColumns = (width: number) => {
		state.screenWidth = width
		if (width >= 1200) {
			state.gridColumns = 12
		}
		else if (width >= 900) {
			state.gridColumns = 3
		}
		else if (width >= 600) {
			state.gridColumns = 2
		}
		else {
			state.gridColumns = 1
		}
	}
	const targetElement = ref<HTMLElement | null>(null)

	onMounted(() => {
		targetElement.value = document.body
		calculateColumns(targetElement.value.clientWidth)
	})
</script>

<template>
	<v-container>
		<v-col>
			<div class="bg-blue-accent-4 top-0 text-center">
				Menu
			</div>
			<div class="page-container">
				<v-col
					v-for="n in 12"
					:key="n"
					class="bg-amber-darken-4"
				>
					{{ n }}
				</v-col>
			</div>
			<div class="bg-blue-accent-4 top-0 text-center">
				Footer
			</div>
		</v-col>
	</v-container>
</template>

<style lang="scss" scoped>
@import '@/assets/tokens.scss';
</style>
