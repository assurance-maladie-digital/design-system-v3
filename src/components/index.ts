export { default as Alert } from './Alert/Alert.vue'
export { default as BackBtn } from './BackBtn/BackBtn.vue'
export { default as BackToTopBtn } from './BackToTopBtn/BackToTopBtn.vue'
export { default as CollapsibleList } from './CollapsibleList/CollapsibleList.vue'
export { default as CopyBtn } from './CopyBtn/CopyBtn.vue'
export { default as CustomInputSelect } from './Customs/CustomInputSelect/CustomInputSelect.vue'
export { default as CustomSelect } from './Customs/CustomSelect/CustomSelect.vue'
export { default as CustomTextField } from './Customs/CustomTextField/CustomTextField.vue'
export { default as DataList } from './DataList/DataList.vue'
export { default as DataListGroup } from './DataListGroup/DataListGroup.vue'
export { default as DownloadBtn } from './DownloadBtn/DownloadBtn.vue'
export { default as FooterBar } from './FooterBar/FooterBar.vue'
export { default as FranceConnectBtn } from './FranceConnectBtn/FranceConnectBtn.vue'
export { default as HeaderBar } from './HeaderBar/HeaderBar.vue'
export { default as HeaderBurgerMenu } from './HeaderBar/HeaderBurgerMenu/HeaderBurgerMenu.vue'
export { default as HeaderMenuItem } from './HeaderBar/HeaderBurgerMenu/HeaderMenuItem/HeaderMenuItem.vue'
export { default as HeaderMenuSection } from './HeaderBar/HeaderBurgerMenu/HeaderMenuSection/HeaderMenuSection.vue'
export { default as HeaderSubMenu } from './HeaderBar/HeaderBurgerMenu/HeaderSubMenu/HeaderSubMenu.vue'
export { default as HeaderLogo } from './HeaderBar/HeaderLogo/HeaderLogo.vue'
export { default as HeaderMenuBtn } from './HeaderBar/HeaderMenuBtn/HeaderMenuBtn.vue'
export { default as HeaderLoading } from './HeaderLoading/HeaderLoading.vue'
export { default as HeaderNavigationBar } from './HeaderNavigationBar/HeaderNavigationBar.vue'
export { default as LangBtn } from './LangBtn/LangBtn.vue'
export { default as Logo } from './Logo/Logo.vue'
export { default as LogoBrandSection } from './LogoBrandSection/LogoBrandSection.vue'
export { default as NotificationBar } from './NotificationBar/NotificationBar.vue'
export * from './NotificationBar/types'
export { default as PageContainer } from './PageContainer/PageContainer.vue'
export { default as SkipLink } from './SkipLink/SkipLink.vue'
export { default as SocialMediaLinks } from './SocialMediaLinks/SocialMediaLinks.vue'
export { default as SubHeader } from './SubHeader/SubHeader.vue'
