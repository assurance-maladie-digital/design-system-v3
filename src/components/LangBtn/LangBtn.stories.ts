import LangBtn from './LangBtn.vue'
import type { Meta, StoryObj } from '@storybook/vue3'
import { VBtn, VMenu, VList, VListItem, VListItemTitle } from 'vuetify/components'

const meta = {
	title: 'Components/LangBtn',
	component: LangBtn,
	argTypes: {
		'modelValue': {
			control: { type: 'select', options: ['fr', 'en', 'es', 'de', '*'] },
			description: 'La langue sélectionnée actuellement',
		},
		'hideDownArrow': {
			control: { type: 'boolean' },
			description: 'Masquer ou afficher la flèche du bouton',
		},
		'ariaLabel': {
			control: { type: 'text' },
			description: 'L\'étiquette ARIA pour l\'accessibilité',
		},
		'availableLanguages': {
			control: { type: 'array' },
			description: 'Les langues disponibles à choisir',
		},
		'update:modelValue': {
			action: 'update:modelValue',
			description: 'Événement émis lors de la sélection de langue',
		},
	},
} as Meta<typeof LangBtn>

export default meta

type Story = StoryObj<typeof meta>

export const Default: Story = {
	args: {
		modelValue: 'en',
		hideDownArrow: false,
		ariaLabel: 'Select a language',
		availableLanguages: ['fr', 'co', 'es'],
	},
	render: (args) => {
		return {
			components: { LangBtn, VBtn, VMenu, VList, VListItem, VListItemTitle },
			setup() {
				return { args }
			},
			template: `
				<div>
					<LangBtn v-bind="args" @update:modelValue="args['update:modelValue']" />
				</div>
			`,
		}
	},
}

export const AllLanguages: Story = {
	args: {
		modelValue: 'fr',
		hideDownArrow: false,
		ariaLabel: 'Select a language',
		availableLanguages: '*',
	},
	render: (args) => {
		return {
			components: { LangBtn, VBtn, VMenu, VList, VListItem, VListItemTitle },
			setup() {
				return { args }
			},
			template: `
				<div>
					<LangBtn v-bind="args" @update:modelValue="args['update:modelValue']" />
				</div>
			`,
		}
	},
}

export const NoDownArrow: Story = {
	args: {
		modelValue: 'fr',
		hideDownArrow: true,
		ariaLabel: 'Select a language',
		availableLanguages: ['fr', 'en', 'de'],
	},
	render: (args) => {
		return {
			components: { LangBtn, VBtn, VMenu, VList, VListItem, VListItemTitle },
			setup() {
				return { args }
			},
			template: `
				<div>
					<LangBtn v-bind="args" @update:modelValue="args['update:modelValue']" />
				</div>
			`,
		}
	},
}
