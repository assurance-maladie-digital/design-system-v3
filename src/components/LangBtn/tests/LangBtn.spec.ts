import { mount } from '@vue/test-utils'
import LangBtn from '../LangBtn.vue'
import { describe, it, expect, afterEach, vi } from 'vitest'
import { createVuetify } from 'vuetify'
import * as components from 'vuetify/components'
import * as directives from 'vuetify/directives'
import ISO6391 from 'iso-639-1'

const vuetify = createVuetify({
	components,
	directives,
})

describe('LangBtn.vue', () => {
	let wrapper

	afterEach(() => {
		if (wrapper) {
			wrapper.unmount()
		}
	})

	it('renders correctly', async () => {
		wrapper = mount(LangBtn, {
			props: {
				availableLanguages: ['fr', 'co', 'es'],
			},
			global: {
				plugins: [vuetify],
			},
		})

		expect(wrapper.html()).toMatchSnapshot()
	})

	it('renders with default props', () => {
		wrapper = mount(LangBtn, {
			props: {
				availableLanguages: ['fr', 'co', 'es'],
			},
			global: {
				plugins: [vuetify],
			},
		})
		expect(wrapper.find('.vd-lang-btn').text()).toBe('Français')
	})

	it('renders with custom modelValue', () => {
		wrapper = mount(LangBtn, {
			props: {
				modelValue: 'co',
				availableLanguages: ['fr', 'co', 'es'],
			},
			global: {
				plugins: [vuetify],
			},
		})
		expect(wrapper.find('.vd-lang-btn').text()).toBe('corsu')
	})

	it('updates selectedLanguage when modelValue prop changes', async () => {
		wrapper = mount(LangBtn, {
			props: {
				modelValue: 'fr',
				availableLanguages: ['fr', 'co', 'es'],
			},
			global: {
				plugins: [vuetify],
			},
		})

		// Vérifier la langue initiale
		expect(wrapper.find('.vd-lang-btn').text()).toBe('Français')

		// Modifier la prop modelValue
		await wrapper.setProps({ modelValue: 'es' })
		await wrapper.vm.$nextTick()

		// Vérifier que la langue affichée a été mise à jour
		expect(wrapper.find('.vd-lang-btn').text()).toBe('Español')
	})

	it('updates language when a language is selected', async () => {
	})

	it('opens the menu when clicked', async () => {
		wrapper = mount(LangBtn, {
			props: {
				availableLanguages: ['fr', 'co', 'es'],
			},
			global: {
				plugins: [vuetify],
			},
			attachTo: document.body,
		})

		// Vérifier que le menu n'est pas ouvert au départ
		expect(document.body.querySelector('.v-list')).toBeNull()

		// Trouver le bouton activator
		const activatorButton = wrapper.find('.vd-lang-btn')

		// Simuler un clic sur le bouton
		await activatorButton.trigger('click')
		await wrapper.vm.$nextTick()

		// Vérifier que le menu est maintenant ouvert
		expect(document.body.querySelector('.v-list')).not.toBeNull()
	})

	it('hides down arrow when hideDownArrow is true', () => {
		wrapper = mount(LangBtn, {
			props: { hideDownArrow: true },
			global: {
				plugins: [vuetify],
			},
		})
		expect(wrapper.find('.v-icon').exists()).toBe(false)
	})

	it('shows down arrow when hideDownArrow is false', () => {
		wrapper = mount(LangBtn, {
			props: { hideDownArrow: false },
			global: {
				plugins: [vuetify],
			},
		})
		expect(wrapper.find('.v-icon').exists()).toBe(true)
	})

	it('renders all languages when availableLanguages is "*"', async () => {
		wrapper = mount(LangBtn, {
			props: {
				availableLanguages: '*',
			},
			global: {
				plugins: [vuetify],
			},
			attachTo: document.body,
		})

		// Ouvrir le menu
		const activatorButton = wrapper.find('.vd-lang-btn')
		await activatorButton.trigger('click')
		await wrapper.vm.$nextTick()

		// Récupérer toutes les langues disponibles via ISO6391
		const allLanguageCodes = ISO6391.getAllCodes()

		// Trouver tous les éléments de la liste
		const listItems = document.body.querySelectorAll('.v-list-item')

		// Vérifier que le nombre d'éléments correspond au nombre de langues
		expect(listItems.length).toBe(allLanguageCodes.length)
	})

	it('renders only specified languages', async () => {
		const languages = ['fr', 'co', 'es']
		wrapper = mount(LangBtn, {
			props: {
				availableLanguages: languages,
			},
			global: {
				plugins: [vuetify],
			},
			attachTo: document.body,
		})

		// Ouvrir le menu
		const activatorButton = wrapper.find('.vd-lang-btn')
		await activatorButton.trigger('click')
		await wrapper.vm.$nextTick()

		// Trouver tous les éléments de la liste
		const listItems = document.body.querySelectorAll('.v-list-item')

		// Vérifier que le nombre d'éléments correspond au nombre de langues spécifiées
		expect(listItems.length).toBe(languages.length)

		// Vérifier que les langues rendues sont correctes
		const renderedLanguages = Array.from(listItems).map(item =>
			item.textContent.trim(),
		)
		const expectedLanguages = languages.map(
			code => ISO6391.getNativeName(code) || code,
		)
		expect(renderedLanguages).toEqual(expectedLanguages)
	})

	it('validates availableLanguages prop correctly with valid values', () => {
		const validator = LangBtn.props.availableLanguages.validator
		expect(validator(['fr', 'en'])).toBe(true)
		expect(validator('*')).toBe(true)
		expect(validator(['es'])).toBe(true)
	})

	it('validates availableLanguages prop correctly with invalid values', () => {
		const validator = LangBtn.props.availableLanguages.validator
		expect(validator([])).toBe(false)
		expect(validator(null)).toBe(false)
		expect(validator(123)).toBe(false)
		expect(validator({})).toBe(false)
		expect(validator('invalid')).toBe(false)
	})

	it('uses default ariaLabel when not provided', () => {
		const defaultAriaLabel = 'Choix de la langue. Actuellement'

		wrapper = mount(LangBtn, {
			props: {
				availableLanguages: ['fr', 'co', 'es'],
			},
			global: {
				plugins: [vuetify],
			},
		})

		const button = wrapper.find('.vd-lang-btn')
		expect(button.attributes('aria-label')).toBe(defaultAriaLabel)
	})

	it('uses ariaLabel prop correctly', () => {
		const customAriaLabel = 'Sélectionnez la langue'
		wrapper = mount(LangBtn, {
			props: {
				ariaLabel: customAriaLabel,
				availableLanguages: ['fr', 'co', 'es'],
			},
			global: {
				plugins: [vuetify],
			},
		})

		const button = wrapper.find('.vd-lang-btn')
		expect(button.attributes('aria-label')).toBe(customAriaLabel)
	})

	it('handles modelValue not in availableLanguages', () => {
		wrapper = mount(LangBtn, {
			props: {
				modelValue: 'unknown',
				availableLanguages: ['fr', 'co', 'es'],
			},
			global: {
				plugins: [vuetify],
			},
		})

		// La langue affichée doit être 'unknown'
		expect(wrapper.find('.vd-lang-btn').text()).toBe('unknown')
	})

	it('falls back to language code when name and nativeName are unavailable', () => {
		// Mock des méthodes ISO6391 pour qu'elles renvoient undefined
		const getNameMock = vi.spyOn(ISO6391, 'getName').mockReturnValue(undefined)
		const getNativeNameMock = vi
			.spyOn(ISO6391, 'getNativeName')
			.mockReturnValue(undefined)

		wrapper = mount(LangBtn, {
			props: {
				modelValue: 'xx', // Code de langue invalide
				availableLanguages: ['xx'],
			},
			global: {
				plugins: [vuetify],
			},
		})

		// La langue affichée doit être le code 'xx'
		expect(wrapper.find('.vd-lang-btn').text()).toBe('xx')

		// Restauration des mocks
		getNameMock.mockRestore()
		getNativeNameMock.mockRestore()
	})
})
