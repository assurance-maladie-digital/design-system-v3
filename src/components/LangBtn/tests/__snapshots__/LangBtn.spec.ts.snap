// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LangBtn.vue > renders correctly 1`] = `
"<div data-v-42678988="" id="lang-menu-8htwqxy0a"><button data-v-42678988="" type="button" class="v-btn v-theme--light text-primary v-btn--density-default v-btn--size-default v-btn--variant-outlined vd-lang-btn" aria-haspopup="menu" aria-controls="lang-menu-8htwqxy0a" aria-owns="v-menu-0" aria-expanded="false"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span>
    <!----><span class="v-btn__content" data-no-activator="">FranÃ§ais <i data-v-42678988="" class="M7,10L12,15L17,10H7Z mdi v-icon notranslate v-theme--light v-icon--size-default ml-1" aria-hidden="true"></i></span>
    <!---->
    <!---->
  </button>
  <!---->
</div>"
`;
